File: src/Tree/BinaryTreePreorderTraversal/BinaryTreePreorderTraversal_Iterative.java
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if(root == null) return res;
        Stack<TreeNode> stack = new Stack<>();

       stack.add(root);

        while(stack.isEmpty() == false){
            TreeNode top = stack.pop();
            //add the current node 
            res.add(top.val);
            if(top.right != null){
                //ensure that right is at the bottom of the left
                stack.add(top.right);
            }
            if(top.left != null){
                //ensure that left is on the top
                stack.add(top.left);
            }
        }
        return res;
    }
}

File: src/Tree/BinaryTreePreorderTraversal/BinaryTreePreorderTraversal_Iterative.py
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res: List[int] = []
        if root is None:
            return res
        stack = []
        
        stack.append(root)
        
        while len(stack) != 0:
            node: TreeNode = stack.pop()
            res.append(node.val)
            
            if node.right is not None:
                stack.append(node.right)
            
            if node.left is not None:
                stack.append(node.left)
        
        return res
